Support for Binary dependencies

UebelAndre opened the pull request:
This change introduces support for `Binary dependencies`. If you take a look at #218, an issue is outlined where `cargo` treats binary dependencies differently than library dependencies and therefore, targets will not be generated when using `cargo raze`. There is now support for a new field in `Cargo.toml` files called `[raze.binaries]`. See the following example:
```toml
[raze.binaries]
wasm-bindgen-cli = "0.2.68"
```

By adding this field and using `genmode = "Remote"` (notably not "Vendored"), Raze will now download the crate from the registry specified in the settings, generate a lock file for it, and add the dependencies and it's metadata to the active raze session and include this metadata in the resulting Bazel files. An example of this can be seen in `//examples/remote/binary_dependencies/...` within this pull request.

This change will also create a `lockfiles` directory in the `workspace_path` for the lockfiles of binary dependencies. The files written will be named `Cargo.{crate}-{version}.lock`. If one of these files exists in the lockfiles directory, whenever `raze` is run again it will inject a lockfile with a matching name given the aforementioned pattern even if the create has a lockfile bundled with it. If no lockfile is present in the lockfiles directory, the bundled lockfile or a newly generated one will be installed there at the end of a run.

```output
...
Generated /cargo-raze/examples/remote/binary_dependencies/cargo/remote/wit-writer-0.2.0.BUILD.bazel successfully
Generated /cargo-raze/examples/remote/binary_dependencies/cargo/BUILD.bazel successfully
Generated /cargo-raze/examples/remote/binary_dependencies/cargo/crates.bzl successfully
Generated /cargo-raze/examples/remote/binary_dependencies/cargo/lockfiles/Cargo.wasm-bindgen-cli-0.2.68.lock successfully  <- Note this output line
Bazel version info
Build label: 3.4.1
...
```

Another example of this change in use can be seen in [bazelbuild/rules_rust#411](https://github.com/bazelbuild/rules_rust/pull/411).

UebelAndre commented:
@damienmg As I mentioned in other pull requests, this is the thing I was working on. Because this change includes the others, it'd be fine to ignore it for now and focus on the others. I also feel this provides functionality similar to #222. I'll need to dig deeper at what the changes are there that were waiting for the recent `crates_index` release.

UebelAndre commented:
Just to note, only the last 3 commits of this PR are the actual change

UebelAndre commented:
@damienmg Ok, this should now only contain the changes relevant to this pull request

damienmg commented:
This is a bit too much details for me and I would be more comfortable if @smklein or @acmcarther would review this one.

UebelAndre commented:
Maybe if there's no movement on [ehuss/cargo-clone-crate#1](https://github.com/ehuss/cargo-clone-crate/pull/1) the code could be incorporated into this project. I would like to try to avoid that if possible but I also don't feel that change will realistically be merged 😞

UebelAndre commented:
[RustSec/cargo-lock#68](https://github.com/RustSec/cargo-lock/issues/68) published `5.0.0-rc` which is better than relying on the git repo. I've updated the dependency to use this version and will update it again once the proper release is out.

UebelAndre commented:
I think I'll need to provide a mechanism to save `Cargo.lock` files of `[raze.binaries]` targets so they can be committed and reused. Right now the change here will download a binary dependency and generate a lockfile for it which is bad for pinning. Maybe `[raze.binaries]` grows a bit to look like
```toml
[raze.binaries]
wasm-bindgen-cli = {lock_file = "path/to/lockfile", version = "0.2.68"}
```

Here if the lockfile doesn't exist `raze` will generate a new one and save it there

Also to note, `version` is the dependency info moved. so it can also look like 
```toml
[raze.binaries]
wasm-bindgen-cli = wasm-bindgen-cli = {lock_file = "path/to/lockfile", version = { version = "0.2.68", features = ["some_feature"]}}
```
Just as another example

UebelAndre commented:
b6b8071b0458cc623acbc3843cf15dca4e96b672, 4fc91f7b0eaa71b4b3806dbabaa8cdc331176349, and bbed53a5b8a3ecabd78c5cd9ad6aae6167a2dd0a have been broken out and put into their own pull requests.

UebelAndre commented:
Moving some of the context from the top description into a comment

This pull request includes #226 #225 and #224 and is the implementation of #218. There are a few outstanding things that should likely be resolved before merging this. See [ehuss/cargo-clone-crate#1](https://github.com/ehuss/cargo-clone-crate/pull/1) and [RustSec/cargo-lock#68](https://github.com/RustSec/cargo-lock/issues/68).

UebelAndre commented:
@damienmg @smklein @acmcarther I think this pull request is ready for review. The only hesitation is that #230 should probably get closed first so 27afe6f14a507c5e558f8d397aad8d2ab072ed19 can be removed from the list of commits.

UebelAndre commented:
@smklein @acmcarther This change is now ready for review.

UebelAndre commented:
@smklein Hey, do you know when you might get a chance to review this? 😅

smklein commented:
Hey, sorry for the delay. I've been occupied with work, which has left me
less time to work on this open source project as I'd like :( I'll see if I
can find some time today.

On Tue, Sep 29, 2020 at 11:39 AM UebelAndre <notifications@github.com>
wrote:

> @smklein <https://github.com/smklein> Hey, do you know when you might get
> a chance to review this? 😅
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/cargo-raze/pull/227#issuecomment-700790948>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAY3T2PNZGNJ5OGDH2IGCFTSIH5TRANCNFSM4RT5MTKQ>
> .
>

UebelAndre commented:
@smklein Thank you so much! I don't mean to pressure you, I was just looking to hear from you and find out if you felt like you had the bandwidth to do a review in the nearish future 😅 

I'm pretty excited about this feature though so I hope the changes are good! 😄

smklein commented on impl/src/planning.rs:
```diff
@@ -355,20 +364,72 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
+        bin_metadatas.push(
+          self
+            .metadata_fetcher
+            .fetch_metadata(&bin_workspace_files, false)
```
This is the only location where `fetch_metadata` avoids using a temporary directory, and it's not totally clear to me *why* it needs this divergent behavior. What's the rationale / what would be different if it used a temporary directory here too?

UebelAndre replied:
`fetch_metadata` fails when I pass the newly fetched source of the binary dependencies. There's no reason to use the `tempfile` logic on these directories because they're already in a temporary directory and there's no way they can poison the repository. Because `fetch_metadata` doesn't handle workspaces well (at all?) it's better to just avoid doing this entirely and running `fetch_metadata` from wherever the binary dependency was cloned.

smklein commented on impl/src/planning.rs:
```diff
@@ -378,30 +439,49 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
 
     let crate_contexts = self.produce_crate_contexts()?;
 
+
```
nit: unnecessary newline

smklein commented on impl/src/planning.rs:
```diff
@@ -378,30 +439,49 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
   pub fn new(metadata_fetcher: &'fetcher mut dyn MetadataFetcher) -> Self {
     Self {
-      metadata_fetcher,
+      metadata_fetcher: metadata_fetcher,
```
```suggestion
      metadata_fetcher,
```

smklein reviewed (commented):
@acmcarther , I still think you should take a look - this only has partial overlap with my contributions to Raze, so I think it would be good to have some of your perspective on "feature-wise, is this something we want", and I can help review the lower-level details.

Top-level comments so far:

1) https://github.com/rust-lang/cargo/issues/8708 implies this binary crates might not be supported long-term by Cargo - specifically, the comment "I don't know if there was ever any intent to support or not support it. Given that it is mostly broken now (and has never worked), a warning would probably help those who stumble on it."

Given this tendency, IMO it makes sense to implement this binary dependency feature in the most non-invasive way possible - I've added some comments regarding refactors for clarity.

2) Can we add some documentation for this feature, possibly in the README, under "handling unconventional crates"?

3) Thanks for the tests!

UebelAndre commented on impl/src/metadata.rs:
```diff
@@ -59,34 +67,153 @@ impl Default for CargoMetadataFetcher {
+fn clone_crate_src(registry_url: &str, dir: &Path, name: &str, version: &str) -> Result<PathBuf> {
+  let url = url::Url::parse(registry_url)?;
+
+  println!("Cloning binary dependency: {}", &name);
```
I wasn't sure if it was valuable information. I kinda like it there but would agree it's kinda verbose. I can remove it.

UebelAndre replied:
Does the update to `dbg!` address everyone's concerns?

UebelAndre replied:
I updated this to use `log::debug!` though I didn't do that for the other print statements. That should be done as a separate PR

smklein commented on impl/src/planning.rs:
```diff
@@ -1054,7 +1174,7 @@ mod checks {
   pub fn warn_unused_settings(
     all_crate_settings: &HashMap<String, CrateSettingsPerVersion>,
-    all_packages: &[Package],
+    all_packages: &Vec<&Package>,
```
BTW, I've seen you make this `&[T]` --> `&Vec<T>` transition in a few places - clippy generally advises [against doing this](https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg), as it prohibits non-vec slices from being used when they might otherwise be valid as an argument.

UebelAndre replied:
The change I was actually trying to make was `&[Package]` -> `&[&Package]`. I didn't know that would work (still learning about the finer details of Rust 😅 

I've updated this. Are there other places? I might be overlooking them.

UebelAndre commented on examples/remote/binary_dependencies/src/main.rs:
```diff
@@ -0,0 +1,38 @@
+
+  // Note: This will need to change if the dependency version changes
+  let wasm_bindgen_path =
+    Path::new("./external/remote_binary_dependencies__wasm_bindgen_cli__0_2_68/cargo_bin_wasm_bindgen");
```
I can't think of an easy way off the top of my head, but would that simplify anything?

UebelAndre commented on impl/src/planning.rs:
```diff
@@ -355,20 +364,72 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
+          &package,
+          &info.req().to_string(),
+        )
+        .unwrap();
```
Updated to use `?`. The use of `unwrap` here was an oversight from an earlier implementation

UebelAndre commented on impl/src/planning.rs:
```diff
@@ -324,14 +331,16 @@ impl CrateCatalog {
 
     let mut package_id_to_entries_idx = HashMap::new();
 
+    // Ensure there are no duplicate packages(?)
```
Thanks for the info

acmcarther commented on impl/src/planning.rs:
```diff
@@ -416,82 +496,122 @@ impl<'planner> WorkspaceSubplanner<'planner> {
-          return None;
-        }
+      .map(|catalog| {
+        (*catalog)
```
Can you refactor this to be more imperative? Create helper functions, use variables, make sure things are clear.

I feel kind of strongly that this is an abomination of function chaining.

(not strongly enough to really push back if you push back, but man this looks gross to me and I think it will be hard for future maintainers to deal with)

UebelAndre replied:
I'm happy you mentioned this because I feel `planning.rs` has kinda grown to be an abomination as well 😬 

I'll gladly look into this but would also love to see this file get split into a module 🤞

UebelAndre replied:
Actually, I would rather leave this and spend some time first splitting `plannint.rs` into a module. I think the lines have become rather blurred between the internal components of this file. I've tried this already and have some good ideas on how to handle the unique situations.

UebelAndre commented:
@smklein Addressed some feedback and asked some questions. Ready for another review 😄

acmcarther commented on impl/src/settings.rs:
```diff
@@ -63,6 +63,12 @@ pub struct RazeSettings {
+   * A list of binary dependencies
+   */
+  #[serde(default)]
+  pub binaries: HashMap<String, cargo_toml::Dependency>,
```
Optional Nit: Not sure that this is self-evident in at usage site inside toml. Maybe would prefer `binary_deps` instead?

I don't feel strongly about this suggestion.

UebelAndre replied:
Updated

acmcarther commented on impl/src/planning.rs:
```diff
@@ -416,82 +496,122 @@ impl<'planner> WorkspaceSubplanner<'planner> {
-      .ok_or_else(|| RazeError::Generic("Missing resolve graph".into()))?
-      .nodes
+    // Additionally, the binary dependencies need to have their checksums added as well
+    for (pkg, info) in self.settings.binaries.iter() {
```
In contrast to comment below, I think this could do to be a bit more functional. I don't feel strongly about it though.

UebelAndre replied:
I think I could use a more direct suggestion here. I struggled a bit writing this section 😅

acmcarther commented on impl/src/planning.rs:
```diff
@@ -416,82 +496,122 @@ impl<'planner> WorkspaceSubplanner<'planner> {
+    // Build a list of all workspace files
+    let mut files: Vec<&CargoWorkspaceFiles> =
+      self.binary_deps_files
+      .iter()
```
Indentation is weird here. Run rustfmt?

(be ready to revert if it causes massive diffs in unchanged code)

UebelAndre replied:
I did run `rustfmt`.  Perhaps the config needs to be updated?

acmcarther commented on impl/src/planning.rs:
```diff
@@ -355,20 +364,72 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
     let crate_catalog = CrateCatalog::new(&metadata)?;
 
-    // Generate additional PlatformDetails
+    // Additionally, fetch metadata for the list of binaries present in raze settings. This
```
Could this bin crates work be done in a helper function, to help keep this function small enough to stay in reader's brain?

UebelAndre replied:
Updated, let me know what you think.

acmcarther reviewed (changes requested):
I got a ton of nits, not really much of substance. I think I've lost the project context needed to produce a real useful comment. I'll make an effort to do so on next pass though.

UebelAndre commented on impl/Cargo.toml:
```diff
@@ -22,23 +22,28 @@ name = "cargo-raze"
-anyhow = "1.0.30"
-cargo_metadata = "0.9.1"
-cargo-lock = "4.0.1"
+anyhow = "1.0.32"
```
Some are, some aren't. I've been doing lots of testing with the dependencies as they are and everything looks fine. If you feel strongly about it I can try to figure out what those dependencies are again to only update what's absolutely required. I wouldn't have touched them if I felt like this project was being used as a library at all but given that the primary use case seems to be as a compiled binary, I think the updated dependencies are good.

acmcarther replied:
No, I don't feel strongly here. Marking Resolved.

UebelAndre commented on impl/src/bazel.rs:
```diff
@@ -505,6 +505,19 @@ impl BuildRenderer for BazelRenderer {
       contents: rendered_bzl_fetch_file,
     });
 
+    // Add all lockfiles for the the binary dependencies
```
1. (unrelated) I feel it's awkward that this file is in `bazel.rs`
2. Anytime `raze` runs with a binary dependency defined, it's going to clone that dependency to determine if anything needs to be updated. However, this may not be desirable in cases where someone reruns `raze` on the same revison months later (where transitive dependencies might have been updated). Writing out and reusing the lockfiles means users can appropriately pin their binary dependencies' transitive dependencies so subsequent runs don't generate new files. Does this answer your question? I can update the comment and the description of this pull request if so.

acmcarther replied:
Most important to me is to capture (2) tersely in the code.

(1) indicates that the separation between the abstractions has broken down (not your fault, probably my fault for not making them clearer)

UebelAndre replied:
I think I've addressed this in my last commit.

UebelAndre commented on impl/src/metadata.rs:
```diff
@@ -59,34 +67,153 @@ impl Default for CargoMetadataFetcher {
+  Ok(bundled_lockfile)
+}
+
+pub fn fetch_crate_src(
```
Happy to move it but I'd like to hear more about why you don't feel this belongs in `metadata`. I ask trying to prevent `util` from becoming a dumping ground for random functionality.

acmcarther replied:
So, caveat on below. I don't think I've constructed a comprehensive mental model for how modules should be delineated in rust, even given the moderate amount of rust code I've written in my life, so everything below is basically "gut feeling", not real principles.

---

In my "platonic ideal world of abstractions", this module is just concerned with the mechanics of loading and extracting metadata about an ensemble of crates. While I do think I would also decide to put functions used by this module which aren't apparently directly related to into it, this function triggered my "this code is tangential" alarm because it was just concerned with fetching remote crates, and it wasn't actually directly used by this module (except in tests). Instead, IIRC, it's used by `planning`. That's a red flag for me.

FWIW, this leads me to a more substantial comment. My intent with the separation of "metadata fetch" and "planning" was to allow all the yucky side-effects like crate fetching and cargo interactions to happen in one black box, and then have "planning" code which was effectively functional over given state in determining what has to get done. Seeing that you're doing things like fetching crates in plan code, we've clearly drifted from that ideal since I originally structured this.

UebelAndre replied:
Thanks for taking the time to share 😄 

While the changes are still in `metadata` right now, I think I've updated the code to better align with your concerns here (though likely still not the ideal state). Is this a step in the right direction?

UebelAndre commented on impl/src/metadata.rs:
```diff
@@ -30,10 +37,11 @@ const SYSTEM_CARGO_BIN_PATH: &str = "cargo";
  */
 pub trait MetadataFetcher {
-  fn fetch_metadata(&mut self, files: &CargoWorkspaceFiles) -> Result<Metadata>;
+  fn fetch_metadata(&mut self, files: &CargoWorkspaceFiles, use_tempdir: bool) -> Result<Metadata>;
```
There isn't an easy way to make this a constructor argument without a larger refactor. I've updated comments around the use of this variable. Even though there are other open comments about `use_tempdir` I'm wondering if this satisfies your request here.

UebelAndre replied:
This turned out to be a tiny refactor 😅 

Updated.

UebelAndre commented:
@acmcarther I primarily have questions for you but also addressed some feedback. I know your time is limited but would greatly appreciate another review so I can get this pull request closed out 😄

acmcarther commented:
Made a really brief pass, but I'll need to sit down and dedicate a block of time to address the other comments (and provide my promised substantive comments).

acmcarther commented:
(In general, despite my grumble grumble about how the abstractions are breaking down, I value new features over satisfying my personal taste in abstraction design, so please don't be too discouraged. As in my prior message, I'll set aside a block of time to get into the nuts and bolts of this PR soon.)

UebelAndre commented:
@acmcarther Thanks! Looking forward to your feedback 😄

UebelAndre commented:
@acmcarther @smklein Hey do you guys have time today to do another review 😅 🙏

UebelAndre commented:
@acmcarther @smklein Hey, friendly ping for another review 😅 

Do you know if you'll have time this week?

UebelAndre commented:
I don't know what to do about the fact that this PR is still flagged as `Changes requested` I think I addressed all I could and had some open questions. I thought maybe resolving the open comments would transition the state of the issue... but today I learned that is not the case 😞  Sorry about that

acmcarther commented:
Sorry for my lack of responsiveness, I operate on Valve time (or Elon Musk time), not human time. Time apparently is very fluid for me. 

You're stuck in the "changes required" stage because you need an explicit LGTM. I think addressing all the requested changes is not sufficient.

Anyway, I haven't gained any new insight in the time I spent being lazy, so I'll just review again (right now) and call it good.

UebelAndre commented:
> Sorry for my lack of responsiveness, I operate on Valve time (or Elon Musk time), not human time. Time apparently is very fluid for me.

Good ol' Elon time 😆 

> You're stuck in the "changes required" stage because you need an explicit LGTM. I think addressing all the requested changes is not sufficient.

Good to know, thanks!


> Anyway, I haven't gained any new insight in the time I spent being lazy, so I'll just review again (right now) and call it good.

@acmcarther You have no idea how excited I am to get these changes through!!! I really hope they're good 🙏 😅

acmcarther reviewed (commented):
I appreciate the effort you've put in here. I'm gonna say some really annoying stuff though.

This is kind of unprecedented for me, but I'm going to insist you split this PR up into more manageable pieces. I've got many issues with the approach here and I'm kind of overwhelmed trying to structure them in a productive way without a tiresome amount of iteration.

Here's a general framework for how this PR could be split up in a way that I could productively review it.

1. *Update the dependencies of cargo-raze*. This will be an easy review and submit.
2. *Implement support for generating metadata in  "in-place".* I'm not certain I'll actually approve this. I might instead ask you to move the responsibility of moving stuff into tempdir outside of the metadata fetcher.
3. *Extend MetadataFetcher to support loading metadata necessary for (future) planning of binary dependencies.* I'm not exactly sure how this should look, but one important point here is that we need additional metadata during planning for binary dependencies that isn't currently available. We should identify what that is and add it in a discrete PR.
4. *Leverage the additional metadata to extend generation to include binary dependencies.*
5. *Add the smoke-tests*.

Some of the particular issues (not covered, or covered obliquely or covered directly above) that motivate this request:
1. The smoke test files being included here are kind of overwhelming.
2. I don't think the planner should be generating files or performing remote requests. The behavior that lives here should probably live in metadata fetcher. That class has precedent for downloading files, network requests, etc. Conceptually, that class should be a "ball of all metadata needed for planning". If planner needs to go get a bunch more stuff, then our abstractions have failed us.
3. I'm not sure how i feel about fields being moved around between WorkspacePlanner, CrateContext, etc. I wanted to address this, but felt a bit overwhelmed by the rest of the PR so I didn't get a chance to devote a lot of thought to it yet.
4. I raised this earlier, but I'm not really comfortable with how the "clone" and "checksum" functions which perform network requests and file operations currently live in metadata, but are primarily used by planning code. With (2), this *is* probably the right place, but right now it feels arbitrary.
5. All together, I feel like these "crate fetching" and "checksum" features deserve a proper abstraction. Something like a trait which represents the "main idea" of what they do, with an implementation to back it. This enables testing of the types that depend on this functionality, without performing the side effects.
6. This is really minor, but we've let the cyclomatic complexity of many of these functions get out of control. In `produce_crate_contexts` there's a `filter_map` with a body that's ~50 lines long, at ~5th level of indentation in the containing function.

UebelAndre commented on impl/src/metadata.rs:
```diff
@@ -12,81 +12,249 @@
 const SYSTEM_CARGO_BIN_PATH: &str = "cargo";
 
+/** A workspace metadata fetcher that uses the Cargo Metadata subcommand. */
+pub struct CargoMetadataFetcher {
```
It seemed to make more sense with previous changes. I've put it back

UebelAndre commented on impl/src/planning.rs:
```diff
@@ -355,20 +364,46 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
+    // Additionally, fetch metadata for the list of binaries present in raze settings. This
+    // is only supported in Remote mode as it's expected that `vendor` has provided all the sources.
+    let (bin_crate_metadata, bin_crate_files) = match settings.genmode {
+      GenMode::Remote => gather_binary_dep_info(
```
I don't think I understand what you're looking for here. I feel like the metadata fetch here is no different than the primary metadata fetch.

acmcarther replied:
I'm giving up on this refactoring request. I hate this, but I really don't care enough to keep pushing.

UebelAndre commented:
>     2\. _Implement support for generating metadata in  "in-place"._ I'm not certain I'll actually approve this. I might instead ask you to move the responsibility of moving stuff into tempdir outside of the metadata fetcher. Basically, your preferred behavior would be come the actual behavior, and the "bonus stuff" around tempdirs would get moved out.

I would like to focus on this a bit more. I don't feel like what I'm doing in this PR is a huge shift in what was being done before but it seems like you feel that it is. I'd like to work on getting on the same page here. Is your concern about `planning.rs` or `metadata.rs`? Either or (if not both), could you elaborate?

UebelAndre commented:
@acmcarther I'm super confused at [your comment](https://github.com/google/cargo-raze/pull/227#pullrequestreview-504359948) It sounds like you're asking for PRs that introduce new (but unused) functionality so that the review is easier? If so, is there no way I can divide commits to make this PR clearer?

UebelAndre commented:
@acmcarther I think one thing that would make things easier is if you could open an issue ticket of some cleanup tasks you would like to see before this PR goes through. This would help me separate issues with the PR and issues in direction (based on my understanding of the existing code).

UebelAndre commented:
@acmcarther I've created #239 #238 #237 #236 #235 to try and reduce the amount of changes in this PR

UebelAndre commented:
@acmcarther Ok, with [those PR's](https://github.com/google/cargo-raze/pull/227#issuecomment-705597712) merged it reduces the footprint of this change a bit more. I think this is now ready to resume discussions in earlier comments.

acmcarther commented:
Can we move smoke tests somewhere else. That's a major gripe I have.

(i recognize that they represent the test coverage for this feature... but maybe we need some unit tests if they're load bearing)

UebelAndre commented:
> Can we move smoke tests and examples somewhere else. That's a major gripe I have.

Removed those commits.

acmcarther commented:
Holy cow you've done it. It's not 20k delta anymore! Thank you very much!

I'm just about out of time for now to review, but I'll be glad to review this in detail ASAP (hopefully later today, but maybe 10AM tomorrow)

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -158,6 +285,117 @@ mod tests {
+    let dir = tempfile::TempDir::new().unwrap();
+    let lockfile_dir = tempfile::TempDir::new().unwrap();
+
+    let (path, _files) = fetch_crate_src(
```
It is important to me that we make sure no unit tests are cloning git repositories. 

This will probably require you to create abstractions which can inject dependencies. Specifically, [cargo_clone::Cloner](https://docs.rs/cargo-clone-crate/0.1.3/cargo_clone/struct.Cloner.html) probably needs to be wrapped and stubbed for testing.

The larger problem here is that you've not really created any abstractions at all.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -63,30 +74,182 @@ impl MetadataFetcher for CargoMetadataFetcher {
+ * lockfile will instead be installed into the crate's directory. If no lockfile exists, one
+ * will be generated.
+ */
+fn generate_lockfile(
```
Nit: `cargo_generate_lockfile`

Make clear that this is shelling out to cargo.

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -63,30 +74,182 @@ impl MetadataFetcher for CargoMetadataFetcher {
   }
 }
 
+fn clone_crate_src(registry_url: &str, dir: &Path, name: &str, version: &str) -> Result<PathBuf> {
```
This functionality should probably be part of the metadata fetcher, not free floating.

acmcarther reviewed (changes requested):
This is really going to take a while

acmcarther commented:
I might just LGTM this after a couple of cycles. Reviewing this PR has made me realize that there is a huge amount of code debt.

UebelAndre commented:
https://github.com/google/cargo-raze/pull/227#discussion_r502587218 What comments?

UebelAndre commented on impl/src/metadata.rs:
```diff
@@ -63,30 +74,182 @@ impl MetadataFetcher for CargoMetadataFetcher {
+  registry_url: &str,
+  lockfiles_dir: &Path,
+  binary_deps_dir: &Path,
+) -> Result<(Vec<Metadata>, HashMap<String, CargoWorkspaceFiles>)> {
```
Can you suggest something you'd prefer?

acmcarther replied:
A type ("dumb" or not) which encodes the semantics of what is being returned.

UebelAndre replied:
Better?

UebelAndre commented:
@acmcarther I don't understand why some comments don't have a field for me to reply to them

UebelAndre commented:
https://github.com/google/cargo-raze/pull/227#discussion_r502582880 I'm gathering metadata in the same place `raze` used to before any of my changes. Are you asking me to make `fetch_metadata` do everything?

UebelAndre commented on impl/src/bazel.rs:
```diff
@@ -505,6 +505,20 @@ impl BuildRenderer for BazelRenderer {
 
+    // In order to support pinned dependencies for binary dependencies, lockfiles are also 
+    // added to ensure the reproducability of subsequent runs.
+    for (name, files) in planned_build.binary_crate_files.iter() {
```
`binary_dependency_lockfiles`

UebelAndre commented:
https://github.com/google/cargo-raze/pull/227#discussion_r502590443 I mocked the endpoints this would attempt to reach out to. This should never be doing any git operations btw. Do you still feel https://docs.rs/cargo-clone-crate/0.1.3/cargo_clone/struct.Cloner.html should be wrapped and mocked?

acmcarther commented on README.md:
```diff
@@ -314,6 +314,24 @@ extra_aliased_targets = [
+Currently, cargo does not gather metadata about crates that do not provide any
+libraries. This means that these specifying them in the `[dependencies]` section
+of your `Cargo.toml` file will not result in generated Bazel targets. Cargo-raze
+has a special feeld to handle these crates when using `genmode = "Remote"`:
```
field

acmcarther commented on impl/src/planning.rs:
```diff
@@ -378,30 +439,49 @@ impl<'fetcher> BuildPlanner for BuildPlannerImpl<'fetcher> {
+      .metadata
+      .packages
+      .iter()
+      .map(|pkg| pkg)
```
???

acmcarther commented on impl/src/planning.rs:
```diff
@@ -416,82 +469,122 @@ impl<'planner> WorkspaceSubplanner<'planner> {
+          .nodes
+          .iter()
+          .sorted_by_key(|n| &n.id)
+          .filter_map(|node| -> Option<Result<CrateContext>> {
```
Move this filter_map body somewhere else or refactor it to be shorter.

UebelAndre replied:
https://github.com/google/cargo-raze/pull/227#discussion_r503446722 
sorry 😅 fixed


https://github.com/google/cargo-raze/pull/227#discussion_r503446934
I added `create_crate_context`

acmcarther commented on impl/src/metadata.rs:
```diff
@@ -63,34 +135,139 @@ impl MetadataFetcher for CargoMetadataFetcher {
+      version,
+      name
+    ))?
+    .1;
```
Do this `.1` in a separate variable declaration.

UebelAndre replied:
Done

acmcarther reviewed (approved):
Well, whatever. If this is idiomatic (which, I think I'm coming to believe it is), then I think I hate modern Rust.

I have lots opinions about things (pattern matching over booleans, ".1" at the end of things, weird arrangement of responsibilities, you including ridiculously huge lambdas in `filter_map`), but I see you're the person that prepared https://github.com/google/cargo-raze/pull/212, so I expect you to be better educated and more invested on the direction of things than I am and for that reason I trust your judgment in terms of where responsibilities should be and what is idiomatic in this language.

Someday I really would like to rewrite this whole thing and fix (what I see as) cruft that has accumulated due to my lack of good discipline in reviewing, but who know when that will happen, and I expect my code would be idiosyncratic w.r.t. the way the rest of the ecosystem writes code.

That all said. please try to prepare much smaller PRs in the future.

UebelAndre commented:
> Well, whatever. If this is idiomatic (which, I think I'm coming to believe it is), then I think I hate modern Rust.
> 
> I have lots opinions about things (pattern matching over booleans, ".1" at the end of things, weird arrangement of responsibilities, you including ridiculously huge lambdas in `filter_map`), but I see you're the person that prepared #212, so I expect you to be better educated and more invested on the direction of things than I am and for that reason I trust your judgment in terms of where responsibilities should be and what is idiomatic in this language.
> 
> Someday I really would like to rewrite this whole thing and fix (what I see as) cruft that has accumulated due to my lack of good discipline in reviewing, but who know when that will happen, and I expect my code would be idiosyncratic w.r.t. the way the rest of the ecosystem writes code.
> 
> That all said. please try to prepare much smaller PRs in the future.

@acmcarther I'm still pretty new to Rust and have learned a lot in contributing to this project. I definitely still have a lot to learn though. I do agree with a lot of your feedback here and I have already made a few branches with some refactoring done that is targeted at creating clear separation of responsibility among the different modules. Though that being said, I'm definitely going to be having a stronger focus on smaller/more focused pull requests. I see a lot of potential in this project and want to do what I can to improve it. Do you think you could open a draft PR or an issue ticket outlining some of the things that bother you most about the current state of the project? I'd be happy to take a look and submit some PRs 😄

UebelAndre commented:
@acmcarther I've squashed the changes in this PR and rebased #240 on top of these changes.

UebelAndre commented:
@acmcarther Thank you so much! Would you mind if I turned that message into an issue and we continued that conversation there? 

Also, is there anything more I should do for this PR?

UebelAndre commented:
I updated the dependencies one last time.

acmcarther commented:
My comment probably represents four separate issues (which may or may not already exist)

1. Improve unit test coverage
2. Replace "smoke-test.sh" with proper functional tests.
3. Evaluate writing crate metadata into `bzl` files for consumption by `rules_rust`
4. Assess lessons from `bazel_gazelle` and produce a proposed set of changes for `cargo-raze`.

acmcarther commented:
Merging now. I'll cut a release that includes this and other recent changes.

UebelAndre commented:
@acmcarther Thank you so so much!!!!

I feel like some of the issues could be cleaned up or consolidated already. It might be worth doing a pass on them 😅

UebelAndre commented:
@acmcarther also, I'd like to use the changes this PR to update `rules_rust` but I have another PR that I've based my other branches off of that I'm hoping to get your eyes on. Could you take a quick look at https://github.com/bazelbuild/rules_rust/pull/420 tomorrow? 🙏 There's one question in-particular I'm hoping to get your eyes on.
